<a href="https://colab.research.google.com/github/vihanga-induwara/Bank-Marketing/blob/main/CM2604.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **1. Load and Import Libraries**

# **2. Load the Dataset**

In [17]:
import pandas as pd


data = pd.read_csv("https://raw.githubusercontent.com/vihanga-induwara/Bank-Marketing/main/bank%2Bmarketing/bank-additional/bank-additional/bank-additional-full.csv", sep=";")

# **3. Explore the Dataset**

## **Inspect the first few rows with .head().**

In [18]:
data.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

## **Check for missing values using .isnull().sum()**

In [19]:
data.isnull().sum()

age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
month             0
day_of_week       0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp.var.rate      0
cons.price.idx    0
cons.conf.idx     0
euribor3m         0
nr.employed       0
y                 0
dtype: int64

## **Check for class imbalance using .value_counts() on the target variable.**

In [20]:
data['y'].value_counts()

y
no     36548
yes     4640
Name: count, dtype: int64

## **Use .describe() for statistical overview of the dataset.**

In [21]:
data.describe()

age      duration      campaign         pdays      previous  \
count  41188.00000  41188.000000  41188.000000  41188.000000  41188.000000   
mean      40.02406    258.285010      2.567593    962.475454      0.172963   
std       10.42125    259.279249      2.770014    186.910907      0.494901   
min       17.00000      0.000000      1.000000      0.000000      0.000000   
25%       32.00000    102.000000      1.000000    999.000000      0.000000   
50%       38.00000    180.000000      2.000000    999.000000      0.000000   
75%       47.00000    319.000000      3.000000    999.000000      0.000000   
max       98.00000   4918.000000     56.000000    999.000000      7.000000   

       emp.var.rate  cons.price.idx  cons.conf.idx     euribor3m   nr.employed  
count  41188.000000    41188.000000   41188.000000  41188.000000  41188.000000  
mean       0.081886       93.575664     -40.502600      3.621291   5167.035911  
std        1.570960        0.578840       4.628198      1.734447     72.251528  
min       -3.400000       92.201000     -50.800000      0.634000   4963.600000  
25%       -1.800000       93.075000     -42.700000      1.344000   5099.100000  
50%        1.100000       93.749000     -41.800000      4.857000   5191.000000  
75%        1.400000       93.994000     -36.400000      4.961000   5228.100000  
max        1.400000       94.767000     -26.900000      5.045000   5228.100000

## **Examine feature types (categorical, continuous) using .dtypes.**

In [22]:
data.dtypes

age                 int64
job                object
marital            object
education          object
default            object
housing            object
loan               object
contact            object
month              object
day_of_week        object
duration            int64
campaign            int64
pdays               int64
previous            int64
poutcome           object
emp.var.rate      float64
cons.price.idx    float64
cons.conf.idx     float64
euribor3m         float64
nr.employed       float64
y                  object
dtype: object

# **4. Data Preprocessing**

4.1 Handle Class Imbalance

4.2 Encode Categorical Variables

4.3 Feature Engineering

4.3.1 Feature Selection:

4.3.2 Feature Extraction (Principal Component Analysis (PCA))

4.5 Normalize/Scale Numerical Features

4.6 Split the Dataset

# **5. Visualize the Data**

5.1 Correlation Heatmap

5.2 Feature Distribution

5.3 Class Distribution

5.4 Feature Relationships

# **6. Train-Test Split**

# **7. Random Forest Model**

7.1 Model Instantiation

7.2 Train the Model

7.3 Evaluate the Model

# **8. Neural Network Model**

8.1 Define the Architecture

8.2 Compile the Model

8.3 Train the Model

8.4 Evaluate the Model

# **9. Hyperparameter Tuning**

9.1 Random Forest Tuning

9.2 Neural Network Tuning

# **10. Model Evaluation**

10.1 Accuracy

10.2 Precision, Recall, F1-Score

10.3 Confusion Matrix

10.4 ROC Curve

# **11. Model Deployment**

Random Forest:

Neural Network: